aten/src/ATen/CPUGenerator.h

```diff
@@ -0,0 +1,43 @@
+#pragma once
+
+#include <ATen/core/Generator.h>
+#include <ATen/core/MT19937RNGEngine.h>
+#include <ATen/core/PhiloxRNGEngine.h>
+
+namespace at {
+
+struct CAFFE2_API CPUGenerator : public CloneableGenerator<CPUGenerator, Generator> {
+  // Constructors
+  CPUGenerator(uint64_t seed_in = default_rng_seed_val);
+  CPUGenerator(mt19937 engine_in);
+
+  // CPUGenerator methods
+  void set_current_seed(uint64_t seed) override;
+  uint64_t current_seed() const override;
+  static DeviceType device_type();
+  uint32_t random();
+  uint64_t random64();
+  uint32_t normal_cache_index();
+  void set_normal_cache_index(uint32_t index);
+  at::detail::Array<float, 2> normal_cache_floats();
+  at::detail::Array<double, 2> normal_cache_doubles();
+  void set_normal_cache_floats(at::detail::Array<float, 2> randoms);
+  void set_normal_cache_doubles(at::detail::Array<double, 2> randoms);
+
+private:
+  CloneableGenerator<CPUGenerator, Generator>* clone_impl() const override;
+  at::mt19937 engine_;
+  uint32_t normal_cache_index_;
+  at::detail::Array<float, 2> normal_cache_floats_;
+  at::detail::Array<double, 2> normal_cache_doubles_;
```

Comment:
I guess these correspond to what were previously:

```
-  /********************************/
-
-  /* For normal distribution */
-  double normal_x;
-  double normal_y;
-  double normal_rho;
-  int normal_is_valid; /* = 0; */
-};
```

Did you read enough about this members to know what they do? If so, it would be greatly appreciated if you could add some docs (I'm asking for an improvement; I am aware the old code was not duplicated). I didn't see any mention of normal handling in your original mega-comment in #13070

Reply:
(Having now read up on Box-Mueller transform, I now know that you get two random numbers per application of the method, so you save up the second random number and give it on the next call instead of generating another pair of random numbers. Saying this explicitly would help a lot! Also, I don't see why you need a `uint32_t` for the cache index; shouldn't a bool be sufficient.)

Reply:
OK, now I understand what you are trying to do... but I think this is going about doing it the wrong way (and the old code was better.) What you have decided to do here, is to save BOTH normal samples you got from B-M. But this is totally pointless; you already gave the user the first normal sample, so you're only ever going to use the second normal sample. In other words, saving the first normal sample is pointless, because it's completely dead forever. Better to just save the second (aka "next") normal sample in this case.